### PR TITLE
[3.6 only] Document transitional feature macros / guards

### LINKED ISSUE #8736: Document compilation guards for stacks
As a developer of Mbed TLS, I want to know which dependencies to put on test cases so that my tests compile in all desired configurations.

For example:

* Can I parse an RSA key with `mbedtls_pk_parse_key`?
* Can I use `mbedtls_pk_sign` with `MBEDTLS_PK_ECDSA` and `MBEDTLS_MD_SHA256`?
* Is SHA-384 supported as the TLS 1.2 PRF algorithm?

This used to be simple: look up the required functions or features (algorithms, TLS extensions, …) in `mbedtls_config.h` (formerly `config.h`). Then we added the PSA API, whose mechanisms are gated by symbols in `include/psa/crypto_config.h`. Still simple. But over the course of Mbed TLS 3.x, through the “driver-only” work, we have grown dual support for two crypto stacks, with two configuration mechanisms. So the question “how do I tell whether algorithm A is supported through function F?” can sometimes be hard to answer. The relevant macros may be defined in `include/*/*adjust*.h` or various `include/mbedtls/*.h`.

The goal of this issue is to document macros such as `MBEDTLS_MD_CAN_SHA256`, `MBEDTLS_ECP_LIGHT`, `MBEDTLS_PK_HAVE_ECC_KEYS`, `MBEDTLS_PK_USE_PSA_EC_DATA`, `MBEDTLS_PK_CAN_ECDSA_VERIFY`, etc. in a way that makes it reasonably easy to go from the intended meaning to the correct macro. Most of the macros are documented where they're defined, but if you don't already know the macro name, that doesn't really help.

This documentation is primarily intended for Mbed TLS maintainers. It would also be useful for maintainers of third-party code that needs to be compatible with all or most Mbed TLS configurations. This documentation is relevant for Mbed TLS 3.x, not 2.28 (no “driver-only” so you just check for legacy algorithms) nor 4.x (no non-PSA configurations so you just check `PSA_WANT`).

### PR DESCRIPTION
**Pull request by mpg**
## Description

Resolves #8736

## PR checklist

- [x] **changelog** not required because: documentation
- [x] **development PR** not required because: 4.0 will complete the transition and have much simpler guards
- [x] **framework PR** not required
- [x] **3.6 PR** here
- [x] **2.28 PR** not required because: 2.28 does not have those guards that were added for driver-only support
- **tests**  not required: documentation

### CODE REVIEW

File: docs/architecture/psa-migration/strategy.md [hunk 1]
```diff
@@ -18,11 +18,15 @@ needs to be changed to use new APIs. For a more detailed account of what's
 implemented, see `docs/use-psa-crypto.md`, where new APIs are about (G2), and
 internal changes implement (G1).
 
-As of early 2023, work towards G5 is in progress: Mbed TLS 3.3 and 3.4 saw
-some improvements in this area, and more will be coming in future releases.
+As of Mbed TLS 3.6 (early 2024, work towards G5 is well advanced: it is now
```
Comment (gilles-peskine-arm) [inline docs/architecture/psa-migration/strategy.md:21]:
```suggestion
As of Mbed TLS 3.6 (early 2024), work towards G5 is well advanced: it is now
```

File: docs/architecture/psa-migration/strategy.md [hunk 2]
```diff
@@ -18,11 +18,15 @@ needs to be changed to use new APIs. For a more detailed account of what's
...
+possible to have hashes/HMAC, ciphers/AEAD, and ECC provided only by drivers,
+with some limitations. See `docs/driver-only-builds.md` for details.
```
Comment (gilles-peskine-arm) [inline docs/architecture/psa-migration/strategy.md:23]:
```suggestion
with some limitations. See [`docs/driver-only-builds.md`](../../driver-only-builds.html) for details.
```

File: docs/architecture/psa-migration/strategy.md [hunk 3]
```diff
@@ -219,7 +223,12 @@ Strategies currently (early 2022) used with each abstraction layer:
 
 - PK (for G1): silently call PSA
 - PK (for G2): opt-in use of PSA (new key type)
-- Cipher (G1): replace calls at each call site
+- PK (for G5): store keys in PSA-friendly format when `ECP_C` is disabled and
+  `USE_PSA` is enabled
+- Cipher (G1, TLS): replace calls at each call site
+- Cipher (G5): create a new internal abstraction layer for (non-DES) block
+  ciphers that silently calls PSA when a driver is available, see
+  `md-cipher-dispatch.md`.
```
Comment (gilles-peskine-arm) [inline docs/architecture/psa-migration/strategy.md:231]:
```suggestion
  [`md-cipher-dispatch.md`](md-cipher-dispatch.html).
```

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Looks correct to me apart from some minor niggles.

However, this does not address https://github.com/Mbed-TLS/mbedtls/issues/8736.

**Comment by mpg** [comment]
> However, this does not address #8736.

Aw, I forgot to add the new file, how embarrassing. Sorry.

File: docs/architecture/psa-migration/transition-guards.md [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+from legacy to PSA in order to determine whether a given cryptographic
+mechanism is available in the current build.
+
+We currently (as of Mbed TLS 3.6) have three sets of feature macros:
+- `PSA_WANT` macros;
+- legacy `MBEDTLS_xxx` macros;
+- transitional `MBEDTLS_xxx` macros that stem from the desire to be able to
+  use crypto mechanisms that are only provided by a driver (G5 in
+`strategy.md`).
+
```
Comment (gilles-peskine-arm) [inline docs/architecture/psa-migration/transition-guards.md:11]: Macros defined in `include/mbedtls/config_adjust_legacy_*.h` only, used outside of `include/mbedtls/*.h` and not mentioned here:

* `MBEDTLS_BLOCK_CIPHER_xxx_VIA_xxx` (used in library, test dependencies)
* `MBEDTLS_BLOCK_CIPHER_SOME_PSA` (used in library, test helpers, test code)
* `MBEDTLS_MD_xxx_VIA_PSA` (used in library, test dependencies)
* `MBEDTLS_PSA_UTIL_HAVE_ECDSA` (used in library, test dependencies)

We should document them as well, but that's not a blocker for this pull request.

**Comment by gilles-peskine-arm** [comment]
This should be on the 3.6 branch, since it won't be relevant in 4.0.

**Comment by mpg** [comment]
@gilles-peskine-arm I added commits addressing your feedback, and then rebased to resolve conflicts and more importantly target the correct branch. Please feel free to review again. Note: I wasn't sure if this is priority medium of high, please feel free to adjust.

**Comment by mpg** [comment]
> @mpg 1. So this shouldn't be a draft PR anymore? 2. The CI isn't happy.

Yeah, I meant to un-draft it but forgot, thanks for the reminder. Regarding the CI, I'm not sure what happened but it's looking happy now.

File: docs/architecture/psa-migration/psa-limitations.md [hunk 5]
```diff
@@ -14,8 +14,8 @@ Limitations relevant for G1 (performing crypto operations)
 Executive summary
 -----------------
 
-- Restartable/interruptible ECC operations: in progress (mid-2024).
-- Arbitrary parameters for FFDH: will be dropped in 4.0.
+- Restartable/interruptible ECC operations: support in PSA will be added in 4.0.
```
Comment (gilles-peskine-arm) [inline docs/architecture/psa-migration/psa-limitations.md:17]: “will be added in 4.0” isn't quite accurate. Interruptible signatures were already in 2.28. Restartable ECDH should be in 4.0. We haven't decided yet about restartable export-public. Their usage in TLS will not be in 4.0.

File: docs/architecture/psa-migration/transition-guards.md [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+is negotiated).
+
+**Pure TLS 1.3 domain:** HMAC is used for the Finished message via PSA Crypto
+APIs. So, TLS 1.3 should depend on `PSA_WANT_ALG_HMAC` - doesn't seem to be
+enforced by `check_config.h` at the moment.
+
+Ciphers (AEAD and unauthenticated)
+----------------------------------
+
+**Overview of existing (internal) APIs:** we currently have 4 (families of)
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply docs/architecture/psa-migration/transition-guards.md:147]: Block cipher modes are still missing herre.

File: docs/architecture/psa-migration/transition-guards.md [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+(`mbedtls_sha256` etc). (That's not true of all example programs, though.)
+
+**Helper macros:** in `config_adjust_legacy_crypto.h` we define a family of
+macro `MBEDTLS_MD_CAN_xxx`. These macros are defined (for available hashes) as
+soon as `MBEDTLS_MD_LIGHT` is enabled. This subset of `MD` is automatically
+enabled as soon as something from the legacy domain, or from the `USE_PSA`
+domain, needs a hash. (Note that this includes `ENTROPY_C`, so in practice
+`MD_LIGHT` is enabled in most builds.)
+
+Note that there is a rule, enforced by `config_adjust_psa_superset_legacy.h`,
```
Comment (gilles-peskine-arm) [inline docs/architecture/psa-migration/transition-guards.md:90]: Not a blocker, but the clarifications I requested on this paragraph (https://github.com/Mbed-TLS/mbedtls/pull/9161/files#r1657124669, https://github.com/Mbed-TLS/mbedtls/pull/9161/files#r1657126157, https://github.com/Mbed-TLS/mbedtls/pull/9161/files#r1657136618https://github.com/Mbed-TLS/mbedtls/pull/9161/files#r1657136618) haven't been addressed.

File: docs/architecture/psa-migration/transition-guards.md [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+- PSA: used by the `USE_PSA` domain when `MBEDTLS_USE_PSA_CRYPTO` is enabled.
+
+**Legacy domain:** most code here is using either `cipher.h` or low-level APIs
+like `aes.h`, and should use legacy macros like `MBEDTLS_AES_C` and
+`MBEDTLS_CIPHER_MOD_CBC`. This includes NIST-KW, CMAC, PKCS5 en/decryption
+functions, PEM decryption, PK parsing of encrypted keys. The only exceptions
+are `GCM` and `CCM` which use the internal abstraction layer `block_cipher`
+and check for availability of block ciphers using `MBEDTLS_CCM_GCM_CAN_xxx`
+macros defined in `config_adjut_legacy_crypto.h`.
+
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply docs/architecture/psa-migration/transition-guards.md:165]: CTR_DRBG is still missing here.

File: docs/architecture/psa-migration/transition-guards.md [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+`MBEDTLS_SSL_HAVE_xxx` macros in `config_adjust_legacy_crypto.h` for the
+ciphers and modes it needs to know about.
+
+**Pure TLS 1.3 domain:** none. All from TLS 1.3 are in the `USE_PSA` domain
+(common to TLS 1.2).
+
+Key derivation
+--------------
+
+**Legacy and `USE_PSA` domains:** no users here.
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply docs/architecture/psa-migration/transition-guards.md:189]: The legacy API does expose key derivation used by code in the USE_PSA domain: PKCS5 and PKCS12.

File: docs/architecture/psa-migration/transition-guards.md [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+is common to ECDH and FFDH thanks to PSA Crypto APIs being generic enough. The
+parts about FFDH are guarded with `PSA_WANT_ALG_FFDH` (with the reasoning that
+this implies support for the corresponding key type).
+
+ECC
+---
+
+**Curves:** in `config_adjut_psa_superset_legacy.h` we ensure that all
+curves that are supported on the legacy side (`MBEDTLS_ECP_DP_xxx_ENABLED`)
+are also supported on the PSA side (`PSA_WANT_ECC_xxx`).
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply docs/architecture/psa-migration/transition-guards.md:225]: As noted before, only if some PSA is included.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Thanks for the updates! This mostly looks good to me, but there's one 4.0 plan that just changed, and a few requests for clarifications from my previous review that you haven't addressed and I think at least some of them should be addressed.

File: docs/architecture/psa-migration/transition-guards.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (mpg) [reply docs/architecture/psa-migration/transition-guards.md:189]: Well, the key derivation functions are exposed but not directly called by library code - PK only uses those key derivations indirectly through the use of the associated password-based encryption scheme. I've added a paragraph anyway.

**Comment by mpg** [comment]
@gilles-peskine-arm Thanks for your review and sorry for forgetting to address some of your comments. I believe I've now addressed all of them except for https://github.com/Mbed-TLS/mbedtls/pull/9161#discussion_r1662777690 which I'd rather leave for later.

File: docs/architecture/psa-migration/transition-guards.md [hunk 11]
```diff
@@ -88,22 +88,22 @@ domain, needs a hash. (Note that this includes `ENTROPY_C`, so in practice
...
+`PSA_WANT_ALG_xxx` and `MBEDTLS_MD_CAN_xxx` are equivalent.
 
 **Legacy and `USE_PSA` domains:** for hashes, `MBEDTLS_MD_CAN_xxx` (where
 `xxx` is the legacy name of the hash) can be used everywhere (except in the
 PSA domain which should use `PSA_WANT` as usual). No special include is
 required, `build_info.h` or `common.h` is enough.
 
 **Pure TLS 1.3 domain:** it is not easy to know which uses of hashes fall in
-this domain as opposed to the `USE_PSA` domain which looking at the code.
+this domain as opposed to the `USE_PSA` domain whithout looking at the code.
```
Comment (gilles-peskine-arm) [inline docs/architecture/psa-migration/transition-guards.md:103]: Trivial:
```suggestion
this domain as opposed to the `USE_PSA` domain without looking at the code.
```

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Review by Harry-Ramsey — APPROVED** [review]
LGTM. I don't have prior experience with 3.6 drivers but this information seems vastly useful for developing ontop of Mbed TLS. The transition guards section will be useful for developing ontop of Mbed TLS or contributing to the library as a whole.

**Pull request merged**